text/0000-variadic-generics.md

```diff
@@ -0,0 +1,209 @@
+- Feature Name: `variadic_generics`
+- Start Date: 2017-2-22
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes the addition of several features to support variadic generics:
+- An intrinsic `Tuple` trait implemented exclusively by tuples
+- `(Head, ...Tail)` syntax for tuple types where `Tail: Tuple`
+- `let (head, ...tail) = tuple;` pattern-matching syntax for tuples
+- `let tuple = (head, ...tail);` syntax for joining an element with a tuple
+
+# Motivation
+[motivation]: #motivation
+
+Variadic generics are a powerful and useful tool commonly requested by Rust
+users (see
+[#376](https://github.com/rust-lang/rfcs/issues/376) and
+[#1921](https://github.com/rust-lang/rfcs/pull/1921)). They allow
+programmers to abstract over non-homogeneous collections, and they make it
+possible to implement functions which accept arguments of varying length and
+type.
+
+Rust has a demonstrable need for variadic generics.
+
+In Rust's own standard library, there are a number of traits which have
+been repeatedly implemented for tuples of varying size up to length 12 using
+macros. This approach has several downsides:
+- It presents arbitrary restrictions on tuples of size 13+.
+- It increases the size of the generated code, resulting in slow compile times.
+- It complicates documentation
+(see the list of trait implementations in
+[this documentation](https://doc.rust-lang.org/std/primitive.tuple.html)).
+
+These arbitrary tuple-length restrictions, manual tuple macros, and confusing
+documentation all combine to increase Rust's learning curve.
+
+Furthermore, community library authors are required to implement similar
+macro-based approaches in order to implement traits for tuples. In the `Diesel`
+crate, it was discovered that replacing macro-generated tuple implementations
+with a structurally-recursive implementation (such as the one proposed here)
+resulted in a 50% decrease in the amount of code generated and a 70% decrease
+in compile times ([link](https://github.com/diesel-rs/diesel/pull/747)). This
+demonstrates that Rust's lack of variadic generics is resulting in a subpar
+edit-compile-debug cycle for at least one prominent, high-quality crate.
+
+The solution proposed here would resolve the limitations above by making it
+possible to implement traits for tuples of arbitrary length. This change would
+make Rust libraries easier to understand and improve the edit-compile-debug
+cycle when using variadic code.
+
+
+# Detailed design
+[design]: #detailed-design
+
+## The `Tuple` Trait
+The following would be implemented by all tuple types:
+```rust
+trait Tuple {
+    type AsRefs<'a>: Tuple + 'a;
+    type AsMuts<'a>: Tuple + 'a;
+    fn elements_as_refs<'a>(&'a self) -> Self::AsRefs<'a>;
+    fn elements_as_mut<'a>(&'a mut self) -> Self::AsMuts<'a>;
```

Comment:
I would prefer to avoid a hard dependency on ATC here, and I'm not sure the functionality needs to be in `Tuple` itself.
IOW, we can do this more transparently to the user with a recursive `impl` perhaps? Well, no, "compiler knows best", it could project `AsRefs` partially (e.g. `(A, ...B, C)::AsRefs == (&A, ...B::AsRefs, &C)`), but we can probably experiment with a recursive `impl` in a library, for a `trait TupleAsRefs<'a>`, to start with.